### PR TITLE
[ocaml] Fix compilation with 4.09.0

### PR DESCRIPTION
**Pull request by ejgallego**
This seems fairly strange, it seems 4.09.0 has somehow modified its
`let` generalization rules?

I will follow up with a PR for adding 4.09 support to the CI [while we discuss this one]

### CODE REVIEW

**Comment by diml** [comment]
That's unexpected indeed. /cc @lpw25 @trefis

File: src/watermarks.ml [hunk 1]
```diff
@@ -28,7 +28,6 @@ let make_watermark_map ~name ~version ~commit =
     match Opam_file.get_field opam_file name with
     | None -> Error (sprintf "variable %S not found in opam file" name)
     | Some value ->
-      let err = Error (sprintf "invalid value for variable %S in opam file" name) in
```
Comment (diml) [inline src/watermarks.ml]: What about `let err () = ...`? That would avoid the duplication

**Comment by rgrinberg** [comment]
octachron said this is a bug with OCaml: https://github.com/ocaml/ocaml/pull/8779

**Comment by lpw25** [comment]
To be clear, it's not really a bug it's a change in unspecified behaviour. Hiding a cmi file is only partially supported and that is what is causing the problem here.

File: src/watermarks.ml (hunk 1 above)
Comment (ejgallego) [reply src/watermarks.ml]: I can do that, but likely we should drop this PR and see if this is fixed upstream.

PR for trunk testing coming soon.

Comment (diml) [reply src/watermarks.ml]: Yep, waiting a little bit to see how it plays out upstream seems good

**Comment by diml** [comment]
To sum up a discussion we had yesterday with @aalekseyev, @lpw25 and @trefis:

Dune is currently providing an `(implicit_transitive_deps false)` feature that is not properly backed up by the compiler. This is not ideal and we should be more careful about such things in the future. Adding proper support for it in the compiler seems easy enough and we can reasonably expect it to land in 4.10. Proper support would take the form of supporting "hidden" includes that could be used by the compiler but not the user. For instance, passing `-Ih <libdir>/core_kernel` (assuming the new flag is called `-Ih`) would allow the compiler to read `core_kernel.cmi` but the user wouldn't be able to use the `Core_kernel` module in their code.

Such a change would make this feature more robust, with the disadvantage that we would loose the speedup provided by the current implementation of `(implicit_transitive_deps false)`. We could recover it to some extent by reducing the number of include flags by reading .cmi files. At this point, it's not clear how hard or if it is even theoretically possible to support the feature as currently implemented in dune, i.e. only passing include flags for dependencies mentioned in the dune file but not for transitive dependencies.

**Comment by diml** [comment]
Let's merge this now. We can always revert it if the compiler patch is merged and included in 4.09. @ejgallego could you sign your commit?

**Comment by rgrinberg** [comment]
I applied Jeremie's fix instead.

**Comment by ejgallego** [comment]
Sounds good, thanks folks!

**Comment by ejgallego** [comment]
Sounds good, thanks folks!

**Comment by ejgallego** [comment]
I'm much afraid that the build is still broken in 4.09, this time in `lib/src.ml` due to some other recent commits :(

**Pull request closed without merging**